pyro/infer/mcmc/nuts.py

```diff
@@ -0,0 +1,135 @@
+from __future__ import absolute_import, division, print_function
+
+from collections import namedtuple
+
+import torch
+
+import pyro
+import pyro.distributions as dist
+from pyro.infer.util import torch_data_sum
+from pyro.ops.integrator import velocity_verlet
+from pyro.util import ng_ones, ng_zeros
+
+from .hmc import HMC
+
+
+_TreeInfo = namedtuple("TreeInfo", ["z_left", "r_left", "z_right", "r_right",
+                                    "z_proposal", "size", "turning", "diverging"])
+
+
+class NUTS(HMC):
+    """
+    No-U-Turn Sampler kernel, where ``step_size`` need to be explicitly specified by the user.
+
+    References
+
+    [1] `The No-U-turn sampler: adaptively setting path lengths in Hamiltonian Monte Carlo`,
+    Matthew D. Hoffman, and Andrew Gelman
+
+    :param model: python callable containing pyro primitives.
+    :param float step_size: determines the size of a single step taken by the
+        verlet integrator while computing the trajectory using Hamiltonian
+        dynamics.
+    """
+
+    def __init__(self, model, step_size=0.5):
+        super(NUTS, self).__init__(model, step_size, num_steps=None)
+        self._reset()
+        self._dE_max = 1000
+
+    def _is_turning(self, z_left, r_left, z_right, r_right):
+        nodes = sorted(z_left)
+        z_left = torch.stack([z_left[name] for name in nodes])
+        r_left = torch.stack([r_left[name] for name in nodes])
+        z_right = torch.stack([z_right[name] for name in nodes])
+        r_right = torch.stack([r_right[name] for name in nodes])
+        dz = z_right - z_left
+        return (torch_data_sum(dz * r_left) < 0) or (torch_data_sum(dz * r_right) < 0)
+
+    def _build_basetree(self, z, r, log_slice):
+        z_new, r_new = velocity_verlet(z, r, self._potential_energy, self.step_size)
+        dE = (log_slice + self._energy(z_new, r_new)).data[0]
+        tree_size = 1 if dE <= 0 else 0
+        diverging = dE >= self._dE_max
+        return _TreeInfo(z_new, r_new, z_new, r_new, z_new, tree_size, False, diverging)
+
+    def _build_tree(self, z, r, log_slice, direction, tree_depth):
```

Comment:
Since this is the complicated bit, can you add a few more comments explaining what's happening?